api/lib/oauth/oauth.js

```diff
@@ -0,0 +1,178 @@
+              }
+
+            } else {
+              var errorObj = {
+                "statusCode": response.statusCode
+              };
+              logger.error("Failed to get space with userId and AppId during permission check", { "appId": appId, "userId": userId, "error": errorObj });
+              callback(errorObj, null);
+            }
+          }
+        });
+      }
+
+
+    });
+
+  }
+
+  function getCloudControllerEndpoint(req) {
+    return settings.cfApi;;
+  };
+
+  function getCloudFoundryInfo(req, callback) {
+    var ccEndpoint = getCloudControllerEndpoint(req);
+    var options = {
+      url: ccEndpoint + "/v2/info",
+      method: "GET",
+      json: true,
+      timeout: 10000,
+
+    };
+    request(options, function(error, response, body) {
+      if (error) {
+        logger.error("Failed to get Cloud Foundry API information", { "info url": options.url, "error": error });
+        error.statusCode = HttpStatus.INTERNAL_SERVER_ERROR;
+        callback(error, null);
+      } else if (response.statusCode === HttpStatus.OK) {
+        logger.info("Get Cloud Foundry API information successfully", { "info url": options.url });
+        callback(null, { "statusCode": HttpStatus.OK, "body": body });
+      } else {
+        var errorObj = {
+          "statusCode": response.statusCode
+        };
+        logger.error("Failed to get Cloud Foundry API information", { "info url": options.url, "error": errorObj });
+        callback(errorObj, null);
+      }
+    });
+  }
+
+
+
+  function getUserInfo(req, callback) {
+    if (obj.authorizationEndpoint) {
+      var userToken = req.header("Authorization");
+      var options = {
+        url: obj.authorizationEndpoint + "/userinfo",
+        method: "GET",
+        json: true,
+        timeout: 10000,
+        headers: {
+          "Authorization": userToken,
+          "Content-Type": "application/json"
+        }
+      };
+      request(options, function(error1, response, body) {
+        if (error1) {
+          logger.error("Failed to get user info from UAA", { "userToken": userToken, "http-options": options, "error": error1 });
+          error1.statusCode = HttpStatus.INTERNAL_SERVER_ERROR;
+          callback(error1, null);
+        } else {
+          if (response.statusCode == HttpStatus.OK) {
+            callback(null, body.user_id);
+          } else {
+            var errorObj = {
+              "statusCode": response.statusCode
+            };
+            logger.error("Failed to get user info from UAA", { "userToken": userToken, "http-options": options, "error": errorObj });
+            callback(errorObj, null);
+          }
+        }
+      });
+    } else {
+      getCloudFoundryInfo(req, function(error, responseBody) {
+        if (error) {
+          callback(error, null);
+        } else {
+          var userToken = req.header("Authorization");
+          obj.authorizationEndpoint = responseBody.body.authorization_endpoint;
+          var options = {
```

Comment:
These codes are duplicated to the implementations of L110-L133.  
How about do checking on obj.authorizationEndpoint first. If not exist, then set it with getCloudFoundryInfo().  Once obj.authorizationEndpoint is set correctly,  invoke /usrinfo directly